### PR TITLE
Get rid of data.holdername

### PR DESCRIPTION
**Pull request by fingolfin**
Temporarily setting and reseting global variables is a dangerous, and can be
avoided here. The result is also arguably easier to understand and debug.

### CODE REVIEW

**Comment by fingolfin** [comment]
@laurentbartholdi could you have a look at my PRs here? I've just fixed a CI test failure in GAP 4.9-4.11 and rebased all PRs, so we'll soon see if tests passed or not.

**Comment by laurentbartholdi** [comment]
@fingolfin I'm not sure I see exactly the improvement: the "data.holdername" trick is theoretically saner, since it seems that you method will fail if a group has a generator called "__the_argument__" (granted, my code will also fail if a group has 26^10 generators consisting of all 10-letter capitalized words; and will even fail randomly...). In all cases, EvalString temporarily creates and destroys a global variable, and furthermore always with the same name; so I'd like to avoid calling EvalString as much as possible.

File: gap/algebra.gi [hunk 1]
```diff
@@ -148,15 +148,8 @@ BindGlobal("TOTALDEGREE@", function(x)
...
 BindGlobal("STRINGSTOLMACHINE@", function(r,args,creator)
-    local temp, i, j, gens, transitions, output, data, Error;
+    local temp, i, j, gens, transitions, output, data;
 
-    Error := function(arg)
-        if IsBound(data) then
-            MakeReadWriteGlobal(data.holdername); Unbind(data.holdername);
-        fi;
-        CallFuncList(VALUE_GLOBAL("Error"),arg);
-    end;
```
Comment (fingolfin) [inline gap/algebra.gi:158]: Another issue with the old code: this trick here to fake installing an exception handler of course only works with `Error` calls made directly in this function. But if any code called by it runs into an error (such as the code being evaluated from a string), then this is not triggered.

**Comment by fingolfin** [comment]
@laurentbartholdi there, now even `__the_argument__` is again an acceptable generator name. I also simplified the code a bit more, merging two arguments to `STRING_WORD2GAP@`.

I hope this can be merged now.

**Comment by laurentbartholdi** [comment]
Very nice, thanks!

**Pull request merged**